fdbserver/TLogGroup.actor.cpp

```diff
@@ -0,0 +1,183 @@
+			// TODO: Currently we add remaining workers to same group. Better way, should just ignore?
+			// TODO: What if there are workers less that GROUP_SIZE?
+			Reference<TLogGroup> group(new TLogGroup());
+			for (const auto& entry : localityMap.getObjects()) {
+				printf("Recruited Group: %s\n", entry->locality.processId().get().toString().c_str());
+				group->addServer(Reference<TLogWorkerData>((TLogWorkerData*)entry));
+				selectedServers.insert(entry->id);
+			}
+			newGroups.push_back(group);
+		}
+	}
+
+	ASSERT_EQ(newGroups.size(), numTeams);
+	recruitedGroups = newGroups;
+}
+
+LocalityMap<TLogWorkerData> TLogGroupCollection::buildLocalityMap(const std::unordered_set<UID>& ignoreServers) {
+	LocalityMap<TLogWorkerData> localityMap;
+	for (const auto& [_, logInterf] : recruitMap) {
+		if (ignoreServers.find(logInterf->id) != ignoreServers.end()) {
+			// Server already selected.
+			continue;
+		}
+		localityMap.add(logInterf->locality, logInterf.getPtr());
+	}
+	return localityMap;
+}
+
+void TLogGroup::addServer(const TLogWorkerDataRef& workerData) {
+	serverMap.emplace(workerData->id, workerData);
+}
+
+std::vector<TLogWorkerDataRef> TLogGroup::servers() const {
+	std::vector<TLogWorkerDataRef> results;
+	for (auto& [_, worker] : serverMap) {
+		results.push_back(worker);
+	}
+	return results;
+}
+
+//-------------------------------------------------------------------------------------------------------------------
+// Unit Tests
+
+namespace testTLogGroup {
+
+// Returns a vector of size 'processCount' containing mocked WorkerInterface, spread across diffeent localities.
+std::vector<WorkerInterface> testTLogGroupRecruits(int processCount) {
+	std::vector<WorkerInterface> recruits;
+	for (int id = 1; id <= processCount; id++) {
+		UID uid(id, 0);
+		WorkerInterface interface;
+		interface.initEndpoints();
+
+		int process_id = id;
+		int dc_id = process_id / 1000;
+		int data_hall_id = process_id / 100;
+		int zone_id = process_id / 10;
+		int machine_id = process_id / 5;
+
+		printf("testMachine: process_id:%d zone_id:%d machine_id:%d ip_addr:%s\n",
+		       process_id,
+		       zone_id,
+		       machine_id,
+		       interface.address().toString().c_str());
+		interface.locality.set(LiteralStringRef("processid"), Standalone<StringRef>(std::to_string(process_id)));
+		interface.locality.set(LiteralStringRef("machineid"), Standalone<StringRef>(std::to_string(machine_id)));
+		interface.locality.set(LiteralStringRef("zoneid"), Standalone<StringRef>(std::to_string(zone_id)));
+		interface.locality.set(LiteralStringRef("data_hall"), Standalone<StringRef>(std::to_string(data_hall_id)));
+		interface.locality.set(LiteralStringRef("dcid"), Standalone<StringRef>(std::to_string(dc_id)));
+		recruits.push_back(interface);
+	}
+	return recruits;
+}
+
+// Checks if each TLog belongs to only one TLogGroup in 'collection', number of workers inside
```

Comment:
I think we want to be flexible: instead of each TLog belongs to only one `TLogGroup`, we prefer that each TLog can belong to multiple `TLogGroup`s so that by randomly assigning SS teams to a TLog group, the load can be balanced.

Reply:
Cool, was actually going to ask that, will also makes it easier for the case where we are out of unique zones for one group -- which is the case now. Thanks!